doc/reference/trd-dynamic-process-loading.md

```diff
@@ -0,0 +1,340 @@
+   indicated via an upcall.
+4. `abort()`: This operation cancels the setup/write operation and frees 
+   allocated resources so that they are available for a different process. 
+   Success or failure is indicated via an upcall.
+
+These operations are implemented using conventional allow, command, and
+subscribe system calls.
+
+
+4 Kernel Capsule Traits
+=================================
+
+The trusted capsules use the following traits to provide functionality to the
+app loading capsule.
+
+4.1 Dynamic Process Binary Storing
+---------------------------------
+
+The kernel provides an interface for storing a process binary loaded at runtime.
+The kernel must expose this interface to ensure that the implementation
+correctly upholds the Tock Threat Model guarantees. Specifically, storing a new
+process binary must not cause existing process binaries to be "lost" or not
+loaded on future reboots of the board.
+
+```rust
+pub trait DynamicBinaryStore {
+    /// Initiate storing a new process binary with the specified length.
+    fn setup(&self, length: usize) -> Result<usize, ErrorCode>;
+
+    /// Store a portion of the process binary.
+    fn write(&self, buffer: SubSliceMut<'static, u8>, offset: usize) -> Result<(), ErrorCode>;
+
+    /// Writing the process binary has finished.
+    fn finalize(&self) -> Result<(), ErrorCode>;
+
+    /// Call to abort the setup/writing process.
+    fn abort(&self) -> Result<(), ErrorCode>;
+
+    fn set_storage_client(&self, client: &'static dyn DynamicBinaryStoreClient);
+}
+
+trait DynamicBinaryStoreClient {
+    /// Any setup work is done and we are ready to write the process binary.
+    fn setup_done(&self);
+
+    /// The provided process binary buffer has been stored.
+    fn write_done(&self, result: Result<(), ErrorCode>, buffer: &'static mut [u8], length: usize);
+
+    /// Operations to finish up writing the new process binary are completed.
+    fn finalize_done(&self, result: Result<(), ErrorCode>);
+
+    /// Canceled any setup or writing operation and freed up reserved space.
+    fn abort_done(&self, result: Result<(), ErrorCode>);
+}
+```
+
+There is a coupling between the `setup()`, `write()`, and `finalize()` calls. The `setup()`
+call allows the implementation to allocate the needed resources to store the
+process binary. Because the kernel requires that it reserves the resources
+required for the new binary before storing it, this method MUST be called before
+calling `write()`. An implementation is responsible for storing individual
+chunks of the process binary.
+Once the entire process binary has been written the `finalize()` method must be
+called. This completes the store and the implementation can handle a new
+`setup()` call.
+
+The `abort()` call deallocates any stored resources and resets the
+implementation to handle a new `setup()` call.
+`abort()` can only be called after `setup()` and before `finalize()`.
```

Comment:
Do we want to expand on what happens when a call to `finalize` fails? Are the resources deallocated, or can we not provide any guarantees in that case?